Add release notes for 0.56.0 (#3294)
microsoft/typespec: docs/release-notes/release-2024-05-07.md

---
title: 0.56 - May 2024
---

# Release Notes Version 0.56 - May 2024

:::warning

> This release contains deprecations and breaking changes

Reviewer: I am missing the deprecations

Author: ah no that was from before when I had the value changes described too

:::

## Breaking Changes

### @typespec/http

> - [#2945](https://github.com/microsoft/typespec/pull/2945) Empty model after removing metadata and visibility property result in void always

Reviewer: ```suggestion
- [#2945](https://github.com/microsoft/typespec/pull/2945) Empty model after removing metadata and applying visibility always results in "void"```

Reviewer 2: Would be good to note the impact this has on OpenAPI emit.

  This means the following case have changed from returning `{}` to no body

  ```tsp
  op b1(): {};
  op b2(): {
    @visibility("none") prop: string;
  };
  op b3(): {
    @added(Versions.v2) prop: string;
  };
  ```

  Workaround: Use explicit `@body`

  ```tsp
  op b1(): {
    @body _: {};
  };
  op b2(): {
    @body _: {
      @visibility("none") prop: string;
    };
  };
  op b3(): {
    @body _: {
      @added(Versions.v2) prop: string;
    };
  };
  ```

- [#2945](https://github.com/microsoft/typespec/pull/2945) Implicit status code always 200 except if response is explicitly `void`

  ```tsp
  op c1(): {
    @header foo: string;
  }; // status code 200 (used to be 204)
  ```

  Solution: Add explicit `@statusCode`

  ```tsp
  op c1(): {
    @header foo: string;
    @statusCode _: 204;
  };
  op c1(): {
    @header foo: string;
    ...NoContent;
  }; // or spread common model
  ```

- [#2945](https://github.com/microsoft/typespec/pull/2945) `@body` means this is the body

>   This change makes it that using `@body` will mean exactly this is the body and everything underneath will be included, including metadata properties. It will log a warning explaining that.

Reviewer 2: ```suggestion
  This change make using `@body` mean this is the exact body and everything underneath will be included, including metadata properties. If metadata properties are present on the body, a warning will be logged.
```

  ```tsp
  op a1(): {@body _: {@header foo: string, other: string} };
                  ^ warning header in a body, it will not be included as a header.
  ```

>   Solution use `@bodyRoot` as the goal is only to change where to resolve the body from.

Reviewer 2: ```suggestion
    Use `@bodyRoot` if you want to only change where to resolve the body from.
```

---
title: 0.56 - May 2024
---

# Release Notes Version 0.56 - May 2024

:::warning
This release contains deprecations and breaking changes
:::

## Values In TypeSpec

> In TypeSpec decorators often expect some sort of configuration to be passed. In version `0.51.0` we introduce the `valueof` keyword to constraint a decorator or template parameter to be a value, however this was limted to string, numeric and boolean values.

Reviewer 2: ```suggestion
Decorators often expect some sort of configuration to be passed as an argument. In version `0.51.0` we introduced the `valueof` keyword to constrain a decorator or template parameter to be a value, however this was limited to string, numeric and boolean values.
```

> When dealing with more complex decorators or free form data one had to use a model type to represent the data. This required a lot of manual validation in the decorator implementation and introduce a lot of potential issues(Passing a subtype like a union instead of a model)

Reviewer 2: ```suggestion
When dealing with more complex decorators or free form data one had to use a model type to represent the data. This required a lot of manual validation in the decorator implementation and introduced a lot of potential for bugs. For example, it was too easy to allow spec authors to pass subtypes that the decorator wouldn't be able to use as a value, like unions.
```
